compiler/rustc_hir_typeck/src/demand.rs

```diff
@@ -983,6 +983,11 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         } else {
             return false;
         }
+        let e_ok = args_e.type_at(0);
+        let f_ok = args_f.type_at(0);
+        if !self.infcx.can_eq(self.param_env, f_ok, e_ok) {
+            return false;
```

Comment:
unrelated to your change but I feel like all of this would read way nicer as one big let chain.

Reply:
Good point, I'll consider doing a little refactor of this function after this PR is merged

Reply:
How's this? https://github.com/rust-lang/rust/pull/116968/commits/24cdb27e2842e5abab2875ce29c365ac5503773e